File: java/robot-simulator/src/main/java/Robot.java
import java.util.Arrays;

class Robot {
    private GridPosition gridPosition;
    private Orientation orientation;

    Robot(GridPosition gridPosition, Orientation orientation) {
        this.gridPosition = gridPosition;
        this.orientation = orientation;
    }

    GridPosition getGridPosition() {
        return this.gridPosition;
    }

    Orientation getOrientation() {
        return this.orientation;
    }

    void turnLeft() {
        orientation = orientation.left();
    }

    void turnRight() {
        orientation = orientation.right();
    }

    void advance() {
        gridPosition = gridPosition.move(orientation);
    }

    void simulate(String steps) {
        Arrays.stream(steps.split("")).forEach(step -> {
            switch (step) {
                case "L": {
                    turnLeft();
                    break;
                }
                case "R": {
                    turnRight();
                    break;
                }
                case "A": {
                    advance();
                    break;
                }
                default:
                    throw new IllegalArgumentException("Invalid step: " + step);
            }
        });
    }
}

File: kotlin/robot-simulator/src/main/kotlin/Robot.kt
class Robot(
    startPosition: GridPosition = GridPosition(0, 0),
    startOrientation: Orientation = Orientation.NORTH
) {
    var gridPosition: GridPosition = startPosition
        private set
    var orientation: Orientation = startOrientation
        private set

    fun advance() {
        gridPosition = gridPosition.move(orientation)
    }
    fun turnLeft() {
        orientation = orientation.left
    }
    fun turnRight() {
        orientation = orientation.right
    }

    fun simulate(move: Char) = when (move) {
        'A' -> advance()
        'L' -> turnLeft()
        'R' -> turnRight()
        else -> error("Invalid move: $move")
    }
    fun simulate(moves: String) {
        for (move in moves) {
            simulate(move)
        }
    }
}
